multiqc/modules/biscuit/biscuit.py

```diff
@@ -0,0 +1,688 @@
+        info="is a tool to map bisulfite converted sequence reads and determine"\
+        " cytosine methylation states.")
+
+        # Set up data structures
+        self.mdata = {
+            'mapq': {},
+            'markdup': {},
+            'coverage': {},
+            'retention': {},
+        }
+
+        # Find and parse bismark alignment reports
+        for k in self.mdata:
+            for f in self.find_log_files('biscuit/%s' % k):
+                d = getattr(self, 'parse_logs_%s' % k)(f['f'], f['fn'])
+                self.mdata[k][f['fn']] = d
+
+        num_parsed = 0
+        for k in self.mdata:
+            num_parsed += len(self.mdata[k])
+        if num_parsed == 0:
+            raise UserWarning
+
+        # Basic Stats Table
+        self.biscuit_stats_table()
+
+        # Write out to the report
+        for k in self.mdata:
+            if len(self.mdata[k]) > 0:
+                self.write_data_file(self.mdata[k], 'multiqc_biscuit_%s' % k)
+                log.info("Found %d biscuit %s reports" % (len(self.mdata[k]), k))
+                if k != 'markdup':
+                    getattr(self, 'chart_biscuit_%s' % k)()
+
+    def biscuit_stats_table(self):
+
+        pd = {}
+        for sid, dd in self.mdata['mapq'].items():
+            if not sid.endswith('_mapq_table.txt'):
+                continue
+            sid = sid.replace('_mapq_table.txt','')
+            allreads = sum([int(_) for _ in dd.values()])
+            pd[sid] = {'%aligned': float(allreads-int(dd['unmapped']))/allreads*100}
+        
+        hasPE = False
+        hasSE = False
+        for sid, dd in self.mdata['markdup'].items():
+            sid = sid.replace('_markdup_report.txt','')
+            sid = sid.replace('_markdup.bam','')
+            sid = sid.replace('.bam','')
+            sid = sid.replace('_markdup','')
+            if sid not in pd:
+                pd[sid] = {}
+            if 'dupRatePE' in dd and dd['dupRatePE'] is not None:
+                pd[sid]['%dupratePE'] = dd['dupRatePE']
+                hasPE = True
+            if 'dupRateSE' in dd and dd['dupRateSE'] is not None:
+                pd[sid]['%duprateSE'] = dd['dupRateSE']
+                hasSE = True
+
+        pheader = OrderedDict()
+        pheader['%aligned'] = {'title':'% Aligned', 'max':100, 'min':0, 'suffix':'%','scale':'Greens'}
+        if hasPE:
+            pheader['%dupratePE'] = {'title':'% DupsPE', 'max':100, 'min':0, 'suffix':'%','scale':'Reds'}
+        if hasSE:
+            pheader['%duprateSE'] = {'title':'% DupsSE', 'max':100, 'min':0, 'suffix':'%','scale':'Reds'}
+        self.general_stats_addcols(pd, pheader)
+
+        # if len(mdata)>0 and all([dd['Duplication Rate'] is not None for dd in mdata.values()]):
+        # self.general_stats_addcols(mdata, {'% Dup': {'suffix':'%', 'min':0, 'max':100}})
+
+
+    def parse_logs_mapq(self, f, fn):
+
+        data = {}
+        if fn.endswith('_mapq_table.txt'):
```

Comment:
Yup, please split each of these out into their own function with their own search pattern.

Reply:
Done.:)